examples/clients_server_select.lua

```diff
@@ -0,0 +1,125 @@
+-- clients_server_select.lua
+local cosock = require "cosock"
+local socket = require "cosock.socket"
+local ip = "0.0.0.0"
+
+local number_of_clients = 10
+
+--- Since the clients and server are in the same application
+--- we can use an OS assigned port and share it across the
+--- two tasks, to coordinate the two tasks to start in the order
+--- we want, we can use a cosock channel to make sure both tasks
+--- have the same port number
+local port_tx, port_rx = cosock.channel.new()
+
+--- Handle a client being ready to recieve
+--- @param client cosock.socket.tcp
+--- @return interger|nil @1 if successful
+--- @return nil|string @nil if successful, error message if not
+function handle_recv(client, clients)
+  local request, err = client:receive()
+  if not request then
+    if err == "closed" then
+      clients[client] = nil
+    end
+    return
+  end
+  print(string.format("receieved %q", request))
+  if request:match("ping") then
+    print("sending pong")
+    local s, err = client:send("pong\n")
+    if err == "closed" then
+      clients[client] = nil
+    elseif err then
+      print("error in recv: " .. tostring(err))
+    end
+  else
+    client:close()
+    clients[client] = nil
+  end
+end
+
+--- Handle a server being ready to accept
+--- @param server cosock.socket.tcp
+--- @return cosock.socket.tcp|nil
+--- @return nil|string @nil if successful, error message if not
+function handle_accept(server, clients)
+  local client, err = server:accept()
+  if err and err ~= "timeout" then
+    error("Error in accept: " .. tostring(err))
+  end
+  clients[client] = true
+end
+
+--- Spawn a task for handling the server side of the socket
+cosock.spawn(function()
+  local server = socket.tcp()
+  server:bind(ip, 0)
+  local _ip, p = server:getsockname()
+  port_tx:send(p)
+  server:listen()
+  local clients = {}
+  handle_accept(server, clients)
+  server:settimeout(0)
+  while true do
+    local recvt = {}
+    for client, _ in pairs(clients) do
+      table.insert(recvt, client)
+    end
+    if #recvt < 5 then
+      table.insert(recvt, server)
+    end
+    local recvr, _sendr, err = cosock.socket.select(recvt, {}, 5)
+    if err == "timeout" then
+      return
```

Comment:
Does it make sense to stop the whole server if there is a `timeout`? It's an example so not a big deal :shrug:

Reply:
Yes, it we go 5 full seconds w/o a new connection we can safely assume all the clients are done so the application is done